pkg/il/program.go

```diff
@@ -0,0 +1,160 @@
+//
+//  - String: represents a string value and maps to the native string Go type. Internally, the
+//  strings are interned in a StringsTable, thus allowing extremely fast equality comparisons. The
+//  id of the string in the StringsTable is encoded in the ByteCode, when stored, or during execution.
+//  - Bool: represents a boolean value and maps to the native bool Go type. The boolean values of
+//  false and true are represented as uint32(0) and uint32(1), respectively. The VM will consider
+//  any non-zero value as true, if it encounters during execution.
+//  - Integer: represents a signed integer value and maps to the native int64 Go type. The integer
+//  is stored in 2 uint32s internally.
+//  - Double: represents a signed float value and maps to the native float64 Go type. The double
+//  is stored in 2 uint32s internally.
+//  - Record represents a mapping of string-> string and maps to the map[string]string Go Type.
+// There is no constant representation of Record in il form.
+//  - Void represents no-value. This is typically used in the return type of a function, to indicate
+// that the function does not return a result.
+//
+// In addition to these types, the VM inherently supports errors, based on Go errors. It is possible
+// to raise these errors from within the IL. However, it is not possible to trap or return these as
+// values. Once an error is raised, whether as a program instruction, or due to an invalid evaluation
+// step, it bubbles all the way out of the evaluation context.
+//
+// The execution model is based on an operand stack and registers. Though most operations work on an
+// operand stack to perform calculations, it is possible to transfer data from registers to stack
+// and vice-versa.
+//
+// The IL supports a stack-based calling-convention. All the parameters to a function needs to be
+// pushed onto the stack before CALL instruction is invoked. Similarly, the return values needs to be
+// pushed onto the stack when RET is called.
+//
+// Typically, an IL based program is created by a compiler by initializing a Program type and
+// by adding Functions to it, whose bodies are built using the Builder type. Once built, programs
+// can be serialized/deserialized into a textual form for ease of use, using Write* and Read* method
+// in this package.
+//
+package il
+
+import (
+	"fmt"
+)
+
+const (
+	// programCodeSize is the default size of a program.
+	programCodeSize int = 256
+)
+
+// Program is a self-contained IL based set of functions that can be executed by an interpreter.
+type Program struct {
+	// Strings is the collection of strings that is used by the program.
+	strings *StringsTable
+
+	// Functions is the collection of functions that are contained in the program.
+	Functions *FunctionsTable
+
+	// Code is the actual byte-code based body of all the functions in the program.
+	code []uint32
+
+	// codeEnd points to the end of the byte-codes within code.
```

Comment:
The make in initialization is allocating a fixed size slice and this is a pointer into it.

But, taking another look, I think it is not "idiomatic". I'll update it to be based on append.